buildSrc/src/main/kotlin/IosPlugin.kt

```diff
@@ -0,0 +1,109 @@
+        val buildBinariesTasks = mutableListOf<String>()
+        target.extensions.configure(KotlinMultiplatformExtension::class.java) {
+            iosArm32(TARGET_NAME_X32) {
+                binaries {
+                    framework()
+                }
+                buildBinariesTasks += compilations.getByName(SourceSet.MAIN_SOURCE_SET_NAME).binariesTaskName
+            }
+            iosArm64(TARGET_NAME_X64) {
+                binaries {
+                    framework()
+                }
+                buildBinariesTasks += compilations.getByName(SourceSet.MAIN_SOURCE_SET_NAME).binariesTaskName
+            }
+            iosX64(TARGET_NAME_SIM) {
+                binaries {
+                    framework()
+                    setupTest(this@iosX64, this, target)
+                }
+                buildBinariesTasks += compilations.getByName(SourceSet.MAIN_SOURCE_SET_NAME).binariesTaskName
+            }
+        }
+        setupBuildAll(target, buildBinariesTasks)
+    }
+
+    private fun setupTest(
+        kotlinNativeTarget: KotlinNativeTarget,
+        kotlinNativeBinaryContainer: KotlinNativeBinaryContainer,
+        target: Project
+    ) {
+        val testBinariesTaskName =
+            kotlinNativeTarget.compilations.getByName(SourceSet.TEST_SOURCE_SET_NAME).binariesTaskName
+        val iosTestProvider = target.tasks.register(TASK_NAME_IOS_TEST, RunIosTestTask::class.java) {
+            group = LifecycleBasePlugin.VERIFICATION_GROUP
+            dependsOn(target.tasks.named(testBinariesTaskName))
+            testExecutables.from(
+                kotlinNativeBinaryContainer.getExecutable(
+                    SourceSet.TEST_SOURCE_SET_NAME,
+                    NativeBuildType.DEBUG
+                ).outputFile
+            )
+        }
+        if (kotlinNativeTarget.publishable) {
+            target.tasks.named(LifecycleBasePlugin.CHECK_TASK_NAME) {
+                dependsOn(iosTestProvider)
+            }
+        }
+    }
+
+    private fun setupBuildAll(
+        target: Project,
+        buildTasks: Collection<String>
+    ) {
+        val allBinariesTaskProvider = target.tasks.register(TASK_NAME_IOS_BINARIES) {
+            dependsOn(buildTasks.map { target.tasks.named(it) })
+        }
+        target.tasks.named(LifecycleBasePlugin.CHECK_TASK_NAME) {
+            dependsOn(allBinariesTaskProvider)
+        }
+    }
+
+    abstract class RunIosTestTask : AbstractTask() {
+
+        @InputFiles
+        open var testExecutables: ConfigurableFileCollection = project.objects.fileCollection()
+
+        @TaskAction
+        open fun runTest() {
+            val files = testExecutables.filter { it.exists() }.files.toTypedArray()
+            if (files.isNotEmpty()) {
+                project.exec {
+                    commandLine("xcrun", "simctl", "spawn", "iPhone X", *files)
+                }
+            } else {
+                logger.error("No test executable for iOS")
```

Comment:
maybe we should throw an exception if it is impossible to run tests?

Reply:
We have modules without tests and I don't know a way to verify if test binary will be created or not. As I know, generate test binary task is still executable even there is no tests. It just generates nothing as output.

Reply:
:(

![giphy](https://user-images.githubusercontent.com/725526/59198800-8126d980-8b8c-11e9-9f08-a0f64115f404.gif)